### DISCUSSION TITLE
Sentence Node (beta)

### DISCUSSION THREAD

**Discussion by zachowj** [zachowj/node-red-contrib-home-assistant-websocket#981] [General]
placeholder

**Comment by matczar** [comment]
Hi,
Is there a way to make a sentence answer other than "done"?

**Reply by zachowj** [reply]
~~Currently, no.~~

**Reply by zachowj** [reply]
This was added to version 0.56.0 of node-red-contrib-home-assistant-websocket and requires version 2.2.0 of hass-node-red

**Reply by matczar** [reply]
Thanks! It was really fast! And I confirm that custom answer works as it should.

**Reply by miGii** [reply]
Can this custom response use wildcard data as variables?

**Reply by donburch888** [reply]
Unfortunately the Response is given _before_ the command can be processed ... so asking "What is the weather" gives response of "done" _before_ the flow is started. And when the flow is finished there seems no way to say the result to the satellite device.

**Comment by dvejsada** [comment]
Would it be possible to have the response as separate node, so it can receive information from the flow? For example, the custom sentence would ask for weather and the response would be formulated using info from weather API?

**Reply by zachowj** [reply]
I think that would be out of the scope of the node. This is more like a custom conversation agent that would get its answers from Node-RED.

https://developers.home-assistant.io/docs/core/conversation/custom_agent

**Reply by danieldioszegi** [reply]
I also liked the idea of having separate responses in different use cases. Like 
- **Turn up the volume** 
- _Turned up the volume_ / _The TV isn't on right now_

It would be also great to be able to provide a list of responses and the node would select one to mimic a "conversation" and not replying always with the same phrase.

Btw I haven't opened NodeRed for months and seeing the sentences node in there was such a huge and pleasent surprise.. Great great work!! :) :)

**Reply by MrAvana** [reply]
Is there a way to implement custom responses that would act like this:

Example config/custom_sentences/en/responses.yaml
language: "en"
responses:
  intents:
    HassTurnOn:
      default: "I have turned on the {{ slots.name }}"

**Reply by skylord123** [reply]
@zachowj Honestly I find it kinda useless without being able to craft custom responses. 99% of the stuff I would use it for in NR will be to ask a question and craft a response. Even if it is to trigger something if that fails I want to be able to tell the end user.

Really hoping this is changed to facilitate that if possible.

**Reply by donburch888** [reply]
I agree that Response should not be part of the sentence node, unless you change the response to "let's see", "i'm thinking" or "ummm". 

The issue is that there is no way to get a message to the satellite device, in order to give a response. 
 Wyoming device and Piper seem to be devices with entities - but neither have any services that can be called.

**Reply by zachowj** [reply]
Custom responses are now available in version [0.72.0](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/releases/tag/v0.72.0)

**Comment by phellarv** [comment]
Is it possible to process variables in the Response-field?
If not, it would be a very nice feature to have.

**Reply by zachowj** [reply]
Custom responses are now available in version [0.72.0](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/releases/tag/v0.72.0)

**Reply by WhoDunDidIt** [reply]
now* ? ;)

Also, thanks so very much!

**Comment by kizovinh** [comment]
Is it possible to add msg.entity to know which entity the command comes from?

**Reply by SpectroHA** [reply]
I came here to ask for this so upvoting here.  

use-case: I want to handle my own responses down stream as service calls, which is easy enough, but I don't have a way right now to know which entity the sentence is coming from (there are 3 possible sources), so I don't know which entity to call for the response.

**Reply by MrAvana** [reply]
Which service call handles responses?

**Reply by SpectroHA** [reply]
just a service call node, domain tts, pointed to a media_player entity.

**Comment by salleq** [comment]
Is there a way to create sentences like in "intents", so for instance the following:
"Tell me [the] weather [in] [backyard|frontyard|outside]" so that you don't have to type each possible sentence separately?

Otherwise this is already a heck lot easier to implement than in HA by restarting every time when you test something!

**Reply by salleq** [reply]
Nevermind, I should've tested it first - it works!

**Comment by donburch888** [comment]
Is there any instruction for getting NR sentence node to be triggered from HA ? 
I followed the example, and have the Call Service node successfully triggering HA Voice Assist; but nothing I have tried triggers the NR Sentence node.

**Reply by zachowj** [reply]
Do you have the Node-RED companion custom integration installed in Home Assistant?

**Reply by donburch888** [reply]
I believe so ... in HACS Integrations 
![image](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/60803198/733962d9-0ccc-4dcf-be12-c993dc5b2b93)
and in node-RED I have
![image](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/60803198/c787c7dd-1a28-4d18-9ff6-bfc835bc9d0e)

**Reply by zachowj** [reply]
Paste in the details from <https://zachowj.github.io/node-red-contrib-home-assistant-websocket/guide/diagnostics.html>

**Reply by donburch888** [reply]
Version: 0.63.1

Home Assistant version: 2024.3.3
Companion version: 0.0.0

Node-RED version: 3.1.7
Docker: yes
Add-on: undefined

Node.js version: v18.19.1 x64 linux
OS: Linux 6.6.20-haos x64

Node.js version: v18.19.1 x64 linux
OS: Linux 6.6.20-haos x64

**Reply by donburch888** [reply]
Companion version should not be 0.0.0.   Should I uninstall and reinstall the companion in HACS ?

**Reply by zachowj** [reply]
Did you complete step 5 <https://github.com/zachowj/hass-node-red?tab=readme-ov-file#hacs>

**Reply by donburch888** [reply]
I guess you are referring to 
> In the Home Assistant UI go to "Configuration" -> "Integrations" click "+" and search for "Node-RED" 

which I think should now read:

From "Settings" in the HA sidebar, select "Devices and Services", click the blue [+ Add integration] button (in bottom right of the page) and search for "node-RED", and install it. 

And the answer is that no, it was not installed.  I have been using NR with HA for a couple of years ... so not sure when that happened. 
The sentence node is now being triggered (though not doing quite what I expected). THANK YOU for your help.

**Comment by WhoDunDidIt** [comment]
Now that HA added [the device_id](https://github.com/home-assistant/core/pull/113094) to the conversation service, it would be amazing if the Sentence node would support it!

**Reply by zachowj** [reply]
This was added in version 0.64.0 and needs the Node-RED Companion, custom component, version 3.1.4+

I was unable to test because only satellite devices return a device id. Someone with a satellite device will have to confirm this is working.

**Reply by WhoDunDidIt** [reply]
Fantastic! I shall test it and let you know!

Thanks so much!

**Reply by WhoDunDidIt** [reply]
Here's my initial test results:

https://i.imgur.com/fuvbqSq.png

**Reply by donburch888** [reply]
Node-RED v3.1.8
node-red-contrib-home-assistant-websocket ... ahh updated from 0.63.1 to 0.64.0
HA node-RED addon v 17.0.11 stopped and restarted
Node-RED companion 3.1.6 ... tick

"Hey Jarvis, what is the weather ?"
*15 second delay*
... and the debug pane shows this slightly before hearing the response "done"
![image](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/60803198/6c674adf-dab5-405a-8f2a-75d2e5552e01)

So yes, it passing one of those hidden internal device IDs.

But how to send the actual calculated result "Current weather is Sunny." back to this device ???
And yes, because the response should be sent back after processing, it is not part of the sentence node.

**Reply by zachowj** [reply]
I am only assuming here but if your satellite device has a HA media player in it you can use the device id as a target of a service call.

**Reply by WhoDunDidIt** [reply]
> I am only assuming here but if your satellite device has a HA media player in it you can use the device id as a target of a service call.

I tested this last night without success.  For now I can just keep a lookup table for device id -> entity id.  Though I wonder if there is a way to just ask HA to provide the entity id for the given device id.

**Comment by donburch888** [comment]
Unfortunately knowing the device_id is useless until the device has a way to play a response.  

Also is the device_id the hidden HA internal ID; or the user-friendly name of the device ?

**Reply by WhoDunDidIt** [reply]
> Unfortunately knowing the device_id is useless until the device has a way to play a response.
> 
> Also is the device_id the hidden HA internal ID; or the user-friendly name of the device ?

My devices CAN play a response though, as long as I know *which* device I should send it to.  I'm using a few Wyoming Satellites running on some Pi's, which have both a mic and a speaker available to it.

My desire is to be able to (while in the kitchen) to say "start a timer for 10 minutes", and have the TTS updates play on the kitchen satellite.  I have the timer and TTS flow ready to go, along with a nifty use of an upper cabinet WLED strip to visualize the timer. 

I'm just waiting for a proper way to determine which satellite picked up the command.

I have a bit of a janky method I'm using now; monitoring each satellites "assist in progress" state and setting a global variable to which device is active. 

It works, but one major downside is that if another satellite receives a assist request, it will change the global variable, so if a long timer is in progress in the kitchen, a request in the bedroom will cause all future timer TTS updates to be moved to the bedroom.

There's plenty of things I could probably do to work around that too, but then it just gets even more janky.

**Reply by skylord123** [reply]
> It works, but one major downside is that if another satellite receives a assist request, it will change the global variable, so if a long timer is in progress in the kitchen, a request in the bedroom will cause all future timer TTS updates to be moved to the bedroom.

When a timer is set you should set the timer_assistant_id or something similar so you output the timer result at the same place it was started at. If you are using something like Node-RED you could easily just define a global variable as an array and add an object to it every time a timer is started with all the details you need. This way the timer result is always returned to the speaker it was started from even if you start two timers from two different speakers.

**Reply by WhoDunDidIt** [reply]
Oh I know there's ways to work around it, but I'd rather avoid layering workaround on top of workaround on top of workaround, when the "proper" solution would be to simply pass the device id that is provided by Home Assistant through to the node.

If I had any idea how to set up a development environment for this project I'd take a stab at it myself, it *should* be fairly trivial.

**Reply by donburch888** [reply]
> My devices CAN play a response though, as long as I know which device I should send it to.

From my research I am guessing that you are sending the TTS to a separate media_player device which just happens to be running on the same physical RasPis (or in the same room) ?? 
If you have found a way to send TTS to the Wyoming satellite, either through Piper or direct, please tell !

**Reply by WhoDunDidIt** [reply]
> From my research I am guessing that you are sending the TTS to a separate media_player device which just happens to be running on the same physical RasPis (or in the same room) ??

You would be correct on this.  However, the Sentence node is able to return a custom response, so it *should* be possible, I just have no clue how you'd do that.

**Reply by donburch888** [reply]
Sorry, but the Response field seems to be plain text.  It is possible to use different text instead of the default "Done", and I have entered a space to make it say nothing. 
And the sentence node happens _before_ I know what response I want to pass back - so the timing is wrong.

**Reply by SpectroHA** [reply]
> > My devices CAN play a response though, as long as I know which device I should send it to.
> 
> From my research I am guessing that you are sending the TTS to a separate media_player device which just happens to be running on the same physical RasPis (or in the same room) ?? If you have found a way to send TTS to the Wyoming satellite, either through Piper or direct, please tell !

The sauce is something like the below in a function node, which is then sent to a call service node.  

A little context on my arrangement.  I have 3 atom echos so I have to determine which echo is activated which is done with a state node for each echo, which writes its entity_id to flow.echo.  If you only have 1 you won't need this and you can just hard code your media player entity id.  After the sentence node and some switch nodes for logic I have a bunch of phrases in a file where a random one is read by an exec node and the phrase is sent to msg.payload, and then handed off to the function node.  Then finally handed to a call service node.

```
var theEcho = flow.get("echo");
var theLanguage;
var theVoice;
var theMessage;
var thePayload;

thePayload = msg.payload.replace(/ /g, '+');
thePayload = thePayload.replace(/[\n\r]/g, '');
theLanguage = "en_US"
theVoice = "en_US-amy-low"
theMessage = "media-source://tts/tts.piper?message=" + thePayload + "&language=" + theLanguage + "&voice=" + theVoice


msg.payload = {
    domain: "media_player", 
    service: "play_media",
    data: {
        entity_id: theEcho,
        media_content_id: theMessage,
        media_content_type: "provider",
        announce: true
    }
}
return msg;
```

**Comment by montaque22** [comment]
Is there a way to filter on the Sentence node by Agent_ID. For example i would like the sentence node to only trigger if a specific agent is used.

**Reply by zachowj** [reply]
Home Assistant processes all sentence triggers, and as far as I know, there’s no built-in way to restrict them to specific `agent_id`s. However, you can filter by `agent_id` after the trigger fires.  

To do this, you’d need to modify the [`handle_trigger`](https://github.com/zachowj/hass-node-red/blob/943d298a67fb99f3cd8ee665339f1fda90d75039/custom_components/nodered/sentence.py#L64) function in the Node-RED companion component to include `agent_id` in the data sent to Node-RED.  

As a quick workaround, I appended `agent_id` to the `result` data so that nothing needs to change on the Node-RED side—just on the Home Assistant side.  

<https://github.com/zachowj/hass-node-red/blob/943d298a67fb99f3cd8ee665339f1fda90d75039/custom_components/nodered/sentence.py#L64-L91>

```python
from homeassistant.components.conversation.models import ConversationInput

...

    @callback
    async def handle_trigger(
        user_input: ConversationInput,
        result: RecognizeResult | None = None,
    ) -> str:
        """
        Handle Sentence trigger.
        RecognizeResult was added in 2023.8.0
        device_id was added in 2024.4.0
        """

        # Convert result to a serializable format
        serialized = convert_recognize_result_to_dict(result)

        # Append agent_id
        serialized["agent_id"] = getattr(user_input, "agent_id", None)

        _LOGGER.debug(f"Sentence trigger: {user_input.text}")
        connection.send_message(
            event_message(
                message_id,
                {
                    "data": {
                        "sentence": user_input.text,
                        "result": serialized,
                        "deviceId": user_input.device_id,
                        "responseId": message_id,
                    }
                },
            )
        )
```

Then, add `"results"` to the output properties of the Sentence node. You can filter downstream using a Switch node, checking if `msg.results.agent_id` matches the desired value.

![image](https://github.com/user-attachments/assets/df6a1fd6-d2f2-4a7a-a686-0c2a940d7537)

**Comment by ITensEI** [comment]
Is there any way to define and react to slots using this node? If I say "Vacuum {room}", the output doesn't contain the room slot anywhere

**Reply by zachowj** [reply]
![image](https://github.com/user-attachments/assets/bec1ee57-ee20-4ceb-bfca-2d43ac80a1a4)

```
[{"id":"be9f3cabbecdd3f0","type":"inject","z":"c89d915bdff0f798","name":"","props":[{"p":"payload"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"living room","payloadType":"str","x":314,"y":1040,"wires":[["a3f2fefff25cc5c4"]]},{"id":"a3f2fefff25cc5c4","type":"api-call-service","z":"c89d915bdff0f798","name":"","server":"","version":7,"debugenabled":false,"action":"conversation.process","floorId":[],"areaId":[],"deviceId":[],"entityId":[],"labelId":[],"data":"{\"text\": \"vacuum \" & payload}","dataType":"jsonata","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","blockInputOverrides":true,"domain":"conversation","service":"process","x":544,"y":1040,"wires":[[]]},{"id":"cd9fbdc6849bbe48","type":"ha-sentence","z":"c89d915bdff0f798","name":"vacuum {room}","server":"2dad33ee.42bf5c","version":2,"inputs":0,"outputs":1,"exposeAsEntityConfig":"","mode":"trigger","sentences":["vacuum {room}"],"response":"","responseType":"str","triggerResponse":"","triggerResponseType":"fixed","responseTimeout":1000,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"results"}],"x":328,"y":1088,"wires":[["699d21b47450b0b8"]]},{"id":"699d21b47450b0b8","type":"debug","z":"c89d915bdff0f798","name":"debug 3","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload.entities","targetType":"msg","statusVal":"","statusType":"auto","x":506,"y":1088,"wires":[]},{"id":"57e75f828f4721a0","type":"inject","z":"c89d915bdff0f798","name":"","props":[{"p":"payload"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"bedroom","payloadType":"str","x":316,"y":992,"wires":[["a3f2fefff25cc5c4"]]}]
```

**Reply by ITensEI** [reply]
Perfect! Thank you!

**Comment by chrispazz** [comment]
Which is the correct way to have a continuous conversation with multiple questions>answers>actions?+

Ex:
User: going out
HA Sentence: should I switch off the lights?
User: yes
HA action to switch off the light
HA Sentence: should I close covers?
User: yes
HA action to close covers

If I use a conversation process after a sentence HA start an asynchronous process and I loose control.

**Reply by zachowj** [reply]
I don’t believe this is currently possible. I haven’t followed the latest voice updates closely, but as far as I know, this kind of feature is only available when using LLMs, not the standard conversation API.

<https://www.home-assistant.io/blog/2025/04/02/release-20254/#continued-conversation-with-llms>
